src/drive/web/modules/public/PublicToolbar.jsx

```diff
@@ -30,68 +30,79 @@ const isShortcutCreatedOnTheRecipientCozy = included => {
   }
   return false
 }
-class PublicToolbar extends React.Component {
-  state = {
-    discoveryLink: null,
-    isSharingShortcutCreated: false
-  }
-
-  componentDidMount() {
-    if (window.location.pathname === '/preview') this.loadSharingDiscoveryLink()
-  }
 
-  async loadSharingDiscoveryLink() {
-    try {
-      const { client } = this.props
-      const response = await client
-        .collection('io.cozy.permissions')
-        .getOwnPermissions()
+const PublicToolbar = ({ hasWriteAccess, refreshFolderContent, files }) => {
+  const [discoveryLink, setDiscoveryLink] = useState()
+  const [isSharingShortcutCreated, setIsSharingSharingcutCreated] = useState(
+    false
+  )
+  const [sharing, setSharing] = useState()
+  const [loading, setLoading] = useState(true)
+  const client = useClient()
 
-      const isSharingShortcutCreated = isShortcutCreatedOnTheRecipientCozy(
-        response.included
-      )
-      const sourceId = response.data.attributes.source_id
-      const sharingId = sourceId.split('/')[1]
-      const { sharecode } = getQueryParameter()
+  useEffect(
+    () => {
+      const loadSharingDiscoveryLink = async () => {
+        setLoading(true)
+        try {
+          const response = await client
+            .collection('io.cozy.permissions')
+            .getOwnPermissions()
 
-      const link = client
-        .collection('io.cozy.sharings')
-        .getDiscoveryLink(sharingId, sharecode)
+          const isSharingShortcutCreated = isShortcutCreatedOnTheRecipientCozy(
+            response.included
+          )
+          const sourceId = response.data.attributes.source_id
+          const sharingId = sourceId.split('/')[1]
+          const { sharecode } = getQueryParameter()
 
-      const { data: sharing } = await client
-        .getStackClient()
-        .fetchJSON('GET', `/sharings/${sharingId}`)
-      this.setState({ discoveryLink: link, isSharingShortcutCreated, sharing })
-    } catch (err) {
-      logger.warn('Failed to load sharing discovery link', err)
-    }
-  }
+          const link = client
+            .collection('io.cozy.sharings')
+            .getDiscoveryLink(sharingId, sharecode)
 
-  render() {
-    const { hasWriteAccess, refreshFolderContent, files } = this.props
-    const { discoveryLink, isSharingShortcutCreated, sharing } = this.state
-    // If there is no DiscoveryLink, it means we're on a sharing by link
-    if (!discoveryLink) {
-      return (
-        <PublicToolbarByLink
-          files={files}
-          hasWriteAccess={hasWriteAccess}
-          refreshFolderContent={refreshFolderContent}
-        />
-      )
-    }
+          const { data: sharing } = await client
+            .getStackClient()
+            .fetchJSON('GET', `/sharings/${sharingId}`)
```

Comment:
Nit: Seems like a lot of business logic here ? Is there a way to extrct the discovery link setup from the component ? Would it make sense ?

Reply:
It's a good idea! a38ded2 and added test